### PR TITLE
fortimanager/fmgr_provisioning.py

### PR DESCRIPTION
**Pull request by Ghilli3**
##### SUMMARY

Adding new module to allow Fortinet customers to perform a 'low touch' deployment of Fortinet devices (primarily FortiGates). This module will allow devices to be pre-configured, so that when the device is deployed, the configuration is automatically pushed down to that device.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

fmgr_provisioning.py
test_fmgr_provisioning.py

##### ANSIBLE VERSION

```
cse@ubuntu:~/PycharmProjects/PR-Provisioning$ ansible --version
ansible 2.5.0 (devel b2282dacdb) last updated 2018/02/05 10:17:29 (GMT -700)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/cse/PycharmProjects/ansible/lib/ansible/modules']
  ansible python module location = /home/cse/PycharmProjects/PR-Provisioning/lib/ansible
  executable location = /home/cse/PycharmProjects/PR-Provisioning/bin/ansible
  python version = 2.7.12 (default, Nov 20 2017, 18:23:56) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/modules/network/fortimanager/fmgr_provisioning.py [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+     is configured, the configuration is pushed down to the registering devices
+
+options:
+  adom:
+    description:
+      - The ADOM the configuration should belong to.
+    required: true
+'''
+
+EXAMPLES = '''
```
Comment (gundalow) [inline lib/ansible/modules/network/fortimanager/fmgr_provisioning.py]: Please give some examples

File: test/units/modules/network/fortimanager/test_fmgr_provisioning.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+# Make coding more python3-ish
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from nose.plugins.skip import SkipTest
+
+try:
```
Comment (gundalow) [inline test/units/modules/network/fortimanager/test_fmgr_provisioning.py:24]: Will need updating to address https://github.com/ansible/ansible/issues/35725

File: lib/ansible/modules/network/fortimanager/fmgr_provisioning.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+            module.fail_json(msg="Installing policy package failed", **results)
+
+        fmg.logout()
+
+        # results is returned as a tuple
+        return module.exit_json(**results[1])
+
+
+if __name__ == "__main__":
+    main()
```
Comment (gundalow) [inline lib/ansible/modules/network/fortimanager/fmgr_provisioning.py]: Missing new line

File: lib/ansible/modules/network/fortimanager/fmgr_provisioning.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+            'sn': serial,
+            'mgmt_mode': 'fmg',
+            'device action': 'add_model',
+             'platform': 'FortiGate-VM64',
+            'os_ver': '5.0',
+            'os_type': 'fos',
+            # 'patch': '0',
+            'desc': 'Provisioned by Ansible',
+        }
+'''
```
Comment (gundalow) [inline lib/ansible/modules/network/fortimanager/fmgr_provisioning.py]: Looks like this is confusing the tests
```
ansible-doc -vvvvv fmgr_provisioning
ansible-doc 2.5.0 (devel 40f7f4babf) last updated 2018/02/05 11:34:06 (GMT +100)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/johnb/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/johnb/git/ansible-inc/ansible-workspace-2/lib/ansible
  executable location = ./bin/ansible-doc
  python version = 3.5.2 (default, Nov 23 2017, 16:37:01) [GCC 5.4.0 20160609]
Using /etc/ansible/ansible.cfg as config file
 [ERROR]: unable to parse /home/johnb/git/ansible-inc/ansible-workspace-2/lib/ansible/modules/network/fmgr_provisioning.py

Traceback (most recent call last):
  File "/home/johnb/git/ansible-inc/ansible-workspace-2/lib/ansible/cli/doc.py", line 163, in run
    doc, plainexamples, returndocs, metadata = get_docstring(filename, fragment_loader, verbose=(self.options.verbosity > 0))
  File "/home/johnb/git/ansible-inc/ansible-workspace-2/lib/ansible/utils/plugin_docs.py", line 116, in get_docstring
    data = read_docstring(filename, verbose=verbose, ignore_errors=ignore_errors)
  File "/home/johnb/git/ansible-inc/ansible-workspace-2/lib/ansible/parsing/plugin_docs.py", line 42, in read_docstring
    M = ast.parse(b_module_data)
  File "/usr/lib/python3.5/ast.py", line 35, in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
  File "<unknown>", line 259
SyntaxError: positional argument follows keyword argument

ERROR! module fmgr_provisioning has a documentation error formatting or is missing documentation.


```

File: lib/ansible/modules/network/fortimanager/fmgr_provisioning.py (hunk 3 above)
Comment (Ghilli3) [reply lib/ansible/modules/network/fortimanager/fmgr_provisioning.py]: Hi @gundalow - I've fixed this and all the other syntax errors.  Thanks

File: test/units/modules/network/fortimanager/test_fmgr_provisioning.py (hunk 2 above)
Comment (Ghilli3) [reply test/units/modules/network/fortimanager/test_fmgr_provisioning.py:24]: Hi @gundalow - I've fixed this issue.  The problem was that I had the case wrong for a python pip package in the test/runner/requirements.txt file.  The docker container was never loading the proper package so all the Fortinet unit tests were failing.

**Comment by Ghilli3** [comment]
I have fixed all the issues so that all the checks have passed and hopefully answered all the review questions.  Please let me know if you have any further questions for me about this module.

Thanks

**Comment by Ghilli3** [comment]
Hi @gundalow - I have a fix for this issue 'Fortinet unit tests is breaking when pyFMG is not installed' which you mentioned in an above comment.

My question is I have two modules so far with Fortinet FortiManager, script and provisioning.  I have fixed the issue with the unit test for provisioning but not with script.  What is the best approach to add a bug fix to the script module?  Shall I create a new pull request and select a 'modification to module'?

Thank you for your time and help

**Comment by gundalow** [comment]
Hi,
We are past the new module cutoff deadline for 2.5.0 now.

PR1: Fix an existing modules & test, under issue type select "bug fix". `@gundalow` in the PR so I see it

Update existing PRs to include this update import logic. They will be merged `devel` and released in Ansible 2.5.0

File: test/units/modules/network/fortimanager/test_fmgr_provisioning.py (hunk 2 above)
Comment (Ghilli3) [reply test/units/modules/network/fortimanager/test_fmgr_provisioning.py:24]: Hi @gundalow  - I have tested and fixed this issue by adding a try/catch block for the import.

**Comment by Ghilli3** [comment]
Hi @gundalow I've created a new PR for the bug fix: https://github.com/ansible/ansible/pull/35917

Currently, the bug fix is included in this PR (fmgr_provisioning) and for (fmgr_script)

Thanks,
Andrew

File: test/units/modules/network/fortimanager/test_fmgr_provisioning.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+# Make coding more python3-ish
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from nose.plugins.skip import SkipTest
```
Comment (maxamillion) [inline test/units/modules/network/fortimanager/test_fmgr_provisioning.py:22]: We'll also need to switch to [pytest](https://docs.pytest.org/en/latest/) for iunit tests as documented [here](https://docs.ansible.com/ansible/devel/dev_guide/testing_units.html#structuring-unit-tests). The Ansible Testing Community Group is attempting to standardize tests moving forward. Thank you!

Comment (gundalow) [reply test/units/modules/network/fortimanager/test_fmgr_provisioning.py:22]: Please raise a follow up PR to fix this once this PR is merged

**Comment by gundalow** [comment]
Merged into `devel` for release in Ansible 2.7.

Can you please raise a follow up PR to move to `pytest`, see the details in https://github.com/ansible/ansible/pull/35743#pullrequestreview-123151841

Thanks in advance

**Comment by Ghilli3** [comment]
@gundalow Sounds good and thanks!

**Pull request merged**